### PR TITLE
Always emit variables from a Swift async context as entry values.

### PR DESCRIPTION
**Pull request by adrian-prantl**
Unfortunately the swiftasync attribute is not yet on LLVM.org.

rdar://70397854

### CODE REVIEW

**Comment by adrian-prantl** [comment]
I'm sure there are a lot of changes in here that can still be upstreamed.

File: llvm/test/DebugInfo/MIR/X86/swift-async-entryvalues.mir [hunk 1]
```diff
@@ -0,0 +1,246 @@
+# RUN: llc -run-pass=livedebugvalues -verify-machineinstrs -march=x86-64 -o - %s | FileCheck %s
+# CHECK: DBG_VALUE renamable $r14, 0, !59, !DIExpression(DW_OP_LLVM_entry_value, 1, DW_OP_deref, DW_OP_plus_uconst, 64, DW_OP_plus_uconst, 24, DW_OP_deref), debug-location !61
+# RUN: llc -O0 -dwarf-version 5 -debugger-tune=lldb -filetype=obj \
+# RUN:   -start-before=livedebugvalues -o - %s \
+# RUN:     | llvm-dwarfdump --name=msg - | FileCheck %s -check-prefixes=CHECK-DWARF
+# CHECK-DWARF:  DW_AT_location	(DW_OP_entry_value(DW_OP_reg14 R14), DW_OP_deref, DW_OP_plus_uconst 0x40, DW_OP_plus_uconst 0x18, DW_OP_deref)
+
```
Comment (vedantk) [inline llvm/test/DebugInfo/MIR/X86/swift-async-entryvalues.mir:22]: Please include steps to regenerate the MIR (perhaps alongside the text of 'async-args.swift').

File: llvm/lib/CodeGen/LiveDebugValues/VarLocBasedImpl.cpp [hunk 2]
```diff
@@ -1811,6 +1818,11 @@ void VarLocBasedLDV::recordEntryValue(const MachineInstr &MI,
   VarLoc EntryValLocAsBackup = VarLoc::CreateEntryBackupLoc(MI, LS, NewExpr);
   LocIndex EntryValLocID = VarLocIDs.insert(EntryValLocAsBackup);
   OpenRanges.insert(EntryValLocID, EntryValLocAsBackup);
+
+  // In Swift async functions entry values are preferred, since they
+  // can be evaluated in both live frames and virtual backtraces.
+  if (isSwiftAsyncContext(MI))
+    MI.getOperand(3).setMetadata(NewExpr);
```
Comment (vedantk) [inline llvm/lib/CodeGen/LiveDebugValues/VarLocBasedImpl.cpp]: I don't think we really want to create `EntryValLocAsBackup` if we're changing the MI's expression (as the backup would be identical to the original VarLoc).

Wdyt of making this change before the following sequence in `ExtendRanges`:

```
  // Only in the case of entry MBB collect DBG_VALUEs representing
  // function parameters in order to generate debug entry values for them.
  MachineBasicBlock &First_MBB = *(MF.begin());
  for (auto &MI : First_MBB) {
    collectRegDefs(MI, DefinedRegs, TRI);
      if (MI.isDebugValue())
        recordEntryValue(MI, DefinedRegs, OpenRanges, VarLocIDs);
  }
```

I.e., rewrite the expressions for all the swift async entry value MI's as the first step.

Then you can skip `recordEntryValue` for swift async MI's entirely.

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

**Comment by adrian-prantl** [comment]
@swift-ci test

**Pull request merged**